src/Base/ContractManager.h

```diff
@@ -0,0 +1,113 @@
+ *                                                                          *
+ *   Copyright (c) 2024 Kacper Donat <kacper@kadet.net>                     *
+ *                                                                          *
+ *   This file is part of FreeCAD.                                          *
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#ifndef CONTRACTMANAGER_H
+#define CONTRACTMANAGER_H
+
+#include <algorithm>
+#include <deque>
+#include <map>
+#include <optional>
+#include <vector>
+#include <list>
+
+namespace Base
+{
+
+class ContractManager
+{
+public:
+    ContractManager() = default;
+
+    template<typename T>
+    T* provide() const
+    {
+        const auto contractName = typeid(T).name();
+
+        if (auto it = _contracts.find(contractName); it != _contracts.end()) {
+            auto contracts = it->second;
+
+            if (contracts.empty()) {
+                return nullptr;
+            }
+
+            return static_cast<T*>(it->second.front());
+        }
+
+        return nullptr;
+    }
+
+    template<typename T>
+    std::list<T*> all() const
+    {
+        if (auto it = _contracts.find(typeid(T).name()); it != _contracts.end()) {
+            auto source = it->second;
+
+            std::list<T*> result(source.size());
+
+            std::transform(source.begin(), source.end(), result.begin(), [](void* ptr) {
+                return static_cast<T*>(ptr);
+            });
+
+            return result;
+        }
+
+        return {};
+    }
+
+    template<typename T>
+    void implement(T* contract)
+    {
+        _contracts[typeid(T).name()].push_front(contract);
+    }
+
+    static ContractManager& get()
+    {
+        if (!_instance) {
```

Comment:
The singleton can be simplified which makes obsolete the source file:
```
static ContractManager& get()
{
    static ContractManager instance;
    return instance;
}
```

Reply:
I had to leave it in the .cpp file as otherwise on some platforms each target would get its own ServiceProvider.